openssl-dynamic/src/main/c/ocsp.c

```diff
@@ -0,0 +1,216 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include "ocsp.h"
+#include "tcn.h"
+#include "ssl_private.h"
+
+#include <stdio.h>
+
+#if !defined(TCN_NO_OCSP) && !defined(OPENSSL_IS_BORINGSSL)
+
+static const int OCSP_CLIENT_ACK = 1;
+static const int OCSP_SERVER_ACK = SSL_TLSEXT_ERR_OK;
+
+/**
+ * This function is being called from OpenSSL. We do everything in
+ * Java-land and this callback is just a stub that returns the
+ * right values to keep OpenSSL happy.
+ *
+ * The arg that is passed into this function is one of these values:
+ *   OCSP_CLIENT_ACK
+ *   OCSP_SERVER_ACK
+ */
+static int openssl_ocsp_callback(SSL *ssl, void *arg) {
+    return *(int*)arg;
+}
+
+#endif /* !defined(TCN_NO_OCSP) && !defined(OPENSSL_IS_BORINGSSL) */
+
+/**
+ * Enables OCSP stapling for the given SSLContext
+ */
+TCN_IMPLEMENT_CALL(jboolean, SSLContext, enableOcsp)(TCN_STDARGS, jlong ctx, jboolean client) {
+#if defined(TCN_NO_OCSP)
+    tcn_ThrowException(e, "OCSP stapling is not supported");
+    return JNI_FALSE;
+
+#elif defined(OPENSSL_IS_BORINGSSL)
+    // This does nothing in BoringSSL
+    return JNI_TRUE;
+
+#else
+    long error = -1L;
+
+    tcn_ssl_ctxt_t *c = J2P(ctx, tcn_ssl_ctxt_t *);
+    if (c == NULL) {
+        tcn_ThrowException(e, "ctx is null");
+        return JNI_FALSE;
+    }
+
+    //
+    // The client and server use slightly different return values to signal
+    // error and success to OpenSSL. We're going to do something naughty to
+    // align OpenSSL's and BoringSSL's APIs and simply tell OpenSSL to use
+    // a stubbed callback function that is always saying that things are OK.
+    //
+    error = SSL_CTX_set_tlsext_status_arg(c->ctx, (client ? &OCSP_CLIENT_ACK : &OCSP_SERVER_ACK));
+    if (error <= 0L) {
+        tcn_ThrowException(e, "SSL_CTX_set_tlsext_status_arg() failed");
+        return JNI_FALSE;
+    }
+
+    error = SSL_CTX_set_tlsext_status_cb(c->ctx, openssl_ocsp_callback);
+    if (error <= 0L) {
+        tcn_ThrowException(e, "SSL_CTX_set_tlsext_status_cb() failed");
+        return JNI_FALSE;
+    }
+
+    return JNI_TRUE;
+#endif
+}
+
+/**
+ * Disables OCSP stapling for the given SSLContext
+ */
+TCN_IMPLEMENT_CALL(void, SSLContext, disableOcsp)(TCN_STDARGS, jlong ctx) {
+// This does nothing in BoringSSL
+#if !defined(TCN_NO_OCSP) && !defined(OPENSSL_IS_BORINGSSL)
+    tcn_ssl_ctxt_t *c = J2P(ctx, tcn_ssl_ctxt_t *);
+    if (c != NULL) {
+      SSL_CTX_set_tlsext_status_cb(c->ctx, NULL);
+      SSL_CTX_set_tlsext_status_arg(c->ctx, NULL);
+    }
+#endif
+}
+
+/**
+ * Enables OCSP stapling on the SSLEngine.
+ */
+TCN_IMPLEMENT_CALL(jboolean, SSL, enableOcsp)(TCN_STDARGS, jlong ssl) {
+#ifdef TCN_NO_OCSP
+  return JNI_FALSE;
```

Comment:
did you consider having this method be `void` and instead throwing an exception upon failure (not supported, `SSL_set_tlsext_status_type` doesn't return `1`? seems less likely to be missed this way.